### PR TITLE
[DO NOT MERGE] Fuse clamp.Tensor backward into a single XPU kernel

### LINKED ISSUE #5155: [release/2.14] torchbench_training_float16 hf_T5_large eager perf regression
### 🐛 Describe the bug

Issue found during 2.14 RC test.

There is performance regression on Windows with Intel XPU (Arc B580). Batch size used: 2

Testcase | Model | Inductor status | Inductor diff | Inductor ref   latency [ms] | Inductor cur   latency [ms] | Eager status | Eager diff | Eager ref   latency [ms] | Eager cur   latency [ms]
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
torchbench_training_float16 | hf_T5_large | PASS | 7.00% | 858.54 | 918.49 | FAIL | 10.70% | 1722.07 | 1907.13

### Versions

<details>
<summary>Click to expand versions</summary>
PyTorch version: 2.14.0+xpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro (10.0.26200 64-bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: N/A

Python version: 3.12.13 | packaged by conda-forge | (main, Aug 11 2026, 10:18:27) [MSC v.1944 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26200-SP0
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: True
XPU used to build PyTorch: 20260100
Intel GPU driver version:
* 32.0.101.8864 (20260717000000.******+***)
Intel GPU models onboard:
* Intel(R) Arc(TM) B580 Graphics
Intel GPU models detected:
* [0] _XpuDeviceProperties(name='Intel(R) Arc(TM) B580 Graphics', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE20B, uuid=86800be2-0000-0000-0300-000000000000, driver_version='1.15.38308', total_memory=11875MB, local_mem_size=128KB, last_level_cache_size=18432KB, max_compute_units=160, memory_clock_rate=0MHz, memory_bus_width=64-bit, gpu_eu_count=160, gpu_subslice_count=20, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1, is_integrated_gpu=0)
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Name: 13th Gen Intel(R) Core(TM) i5-13400
Manufacturer: GenuineIntel
Family: 205
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 2500
MaxClockSpeed: 2500
L2CacheSize: 9728
L2CacheSpeed: None
Revision: None

Versions of relevant libraries:
[pip3] dpcpp-cpp-rt==2026.1.0
[pip3] intel-cmplr-lib-rt==2026.1.0
[pip3] intel-cmplr-lib-ur==2026.1.0
[pip3] intel-cmplr-lic-rt==2026.1.0
[pip3] intel-opencl-rt==2026.1.0
[pip3] intel-openmp==2026.1.0
[pip3] intel-pti==1.0.1
[pip3] intel-sycl-rt==2026.1.0
[pip3] mkl==2026.1.0
[pip3] mkl-include==2024.2.0
[pip3] mkl-static==2024.2.0
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.26.4
[pip3] onemkl-license==2026.1.0
[pip3] onemkl-sycl-blas==2026.1.0
[pip3] onemkl-sycl-dft==2026.1.0
[pip3] onemkl-sycl-lapack==2026.1.0
[pip3] onemkl-sycl-rng==2026.1.0
[pip3] onemkl-sycl-sparse==2026.1.0
[pip3] onnx==1.21.0
[pip3] onnx-ir==0.1.16
[pip3] onnxscript==0.6.2
[pip3] optree==0.13.0
[pip3] tbb==2023.1.0
[pip3] tbb-devel==2021.13.1
[pip3] tcmlib==1.5.0
[pip3] torch==2.14.0+xpu
[pip3] torchaudio==2.11.0+xpu
[pip3] torchvision==0.29.0+xpu
[pip3] triton-xpu==3.8.0
[pip3] umf==1.1.0
[conda] dpcpp-cpp-rt                 2026.1.0         pypi_0              pypi
[conda] intel-cmplr-lib-rt           2026.1.0         pypi_0              pypi
[conda] intel-cmplr-lib-ur           2026.1.0         pypi_0              pypi
[conda] intel-cmplr-lic-rt           2026.1.0         pypi_0              pypi
[conda] intel-opencl-rt              2026.1.0         pypi_0              pypi
[conda] intel-openmp                 2026.1.0         pypi_0              pypi
[conda] intel-pti                    1.0.1            pypi_0              pypi
[conda] intel-sycl-rt                2026.1.0         pypi_0              pypi
[conda] mkl                          2026.1.0         pypi_0              pypi
[conda] mkl-include                  2024.2.0         pypi_0              pypi
[conda] mkl-static                   2024.2.0         pypi_0              pypi
[conda] numpy                        1.26.4           pypi_0              pypi
[conda] onemkl-license               2026.1.0         pypi_0              pypi
[conda] onemkl-sycl-blas             2026.1.0         pypi_0              pypi
[conda] onemkl-sycl-dft              2026.1.0         pypi_0              pypi
[conda] onemkl-sycl-lapack           2026.1.0         pypi_0              pypi
[conda] onemkl-sycl-rng              2026.1.0         pypi_0              pypi
[conda] onemkl-sycl-sparse           2026.1.0         pypi_0              pypi
[conda] optree                       0.13.0           pypi_0              pypi
[conda] tbb                          2023.1.0         pypi_0              pypi
[conda] tbb-devel                    2021.13.1        pypi_0              pypi
[conda] tcmlib                       1.5.0            pypi_0              pypi
[conda] torch                        2.14.0+xpu       pypi_0              pypi
[conda] torchaudio                   2.11.0+xpu       pypi_0              pypi
[conda] torchvision                  0.29.0+xpu       pypi_0              pypi
[conda] triton-xpu                   3.8.0            pypi_0              pypi
[conda] umf                          1.1.0            pypi_0              pypi

</details>

### PR DESCRIPTION
**Pull request by PatrykWilczewski**
- Problem: float16 training of the hf_T5_large model regressed on Arc B580. The two-sided clamp.Tensor backward runs as an 11-op element-wise composite, and the model issues 120 such calls per training step, independent of batch size. The composite's cost is nearly shape-independent from 8 up to 1e6 elements, meaning it is dominated by per-kernel launch overhead rather than by memory bandwidth, so the number of ops dispatched is what drives the cost.
- Root cause: pytorch/pytorch#191142 grew the two-sided clamp_backward(grad, self, Tensor min, Tensor max) formula from 5 to 11 element-wise ops so that the 1/2-gradient is split correctly at ties. The enw semantics are intentional and documented (minimum-norm subradient, 1/2 to each bound at a tie when lo < hi), so the correct response is to fuse the ops, not to revert them.
- Fix: implement _clamp_backward_tensor_xpu for the private op introduced by the corresponding PyTorch change.
  - ClampBackwardFunctor<scalar_t> applies the three-region rule in a
    single gpu_kernel pass: inactive -> 0, tie -> grad/2, interior ->
    grad. NaN comparisons are always false, which preserves NaN pass-through.
  - TensorIteratorConfig uses promote_inputs_to_common_dtype(true) so the bounds are compared in the promoted dtype instead of being narrowed to the activation dtype.
- Validation:
  - 68/68 semantics checks pass, bit-exact (atol=0) against both the 11-op composite on XPU and the same aten op on CPU: boundary cases including equal, reversed and reversed-at-min bounds and NaN self; float16, bfloat16, float32 and float64; mixed and reverse-mixed bound dtypes; a bound not representable in the activation dtype; broadcast 0-dim and row bounds; non-contiguous inputs; 4096 randomized elements per dtype in two dtypes, drawn so that exact tie are hit; gradcheck adn gradgradcheck
  - pytest test/test_binary_ufuncs.py -k "clamp and xpu": 486 passed
  - pytest test/test_ops.py -k "clamp and xpu": 90 passed
  - pytest test/xpu/test_binary_ufuncs_xpu.py -k clamp: 486 passed
  - microbenchmark of the backward op alone, at the shape and dtypes hf_T5_large actually uses ([2,128,1024] float16 activations, 0-dim float32 bounds), taking the current 11-op composite as the baseline:
      * the 5-op formula from before #191142: 79% faster than the baseline
      * this kernel: 86% faster than the baseline
  - end-to-end hf_T5_large float16 forward and backward: the step is 7.6% faster at batch size 1 and 7.5% faster at batch size 2.

Partially fixes: intel/torch-xpu-ops#5155
Recovers 16% of the reproted regression. The remaining has a different cause that is still being investigated.

### CODE REVIEW

**Comment by PatrykWilczewski** [comment]
Do not merge before https://github.com/pytorch/pytorch/pull/196001
